### PR TITLE
Export FileWidget as component rather than type

### PR DESCRIPTION
**Pull request by otaviojacobi**
Change-type: patch

### CODE REVIEW

File: src/components/Form/Widgets/FileWidget.tsx [hunk 1]
```diff
@@ -204,9 +207,11 @@ export const FileWidget = ({
 							sx={{ fontSize: 64, color: (theme) => theme.palette.grey[500] }}
 						/>
 					)}
-					<Typography ml={2}>Drag and drop or browse files</Typography>
-				</Box>
-			</Box>
+					<Typography ml={2}>
```
Comment (myarmolinsky) [inline src/components/Form/Widgets/FileWidget.tsx:210]: remove `ml={2}` for `gap={2}` on the Stack

Comment (otaviojacobi) [reply src/components/Form/Widgets/FileWidget.tsx:210]: Are you doing it or do you want me to push :P ?

File: src/components/Form/Widgets/FileWidget.tsx [hunk 2]
```diff
@@ -1,248 +1,429 @@
...
-
-		if ('dataUrl' in dataUrlValue) {
-			return dataUrlValue.dataUrl;
-		}
-	};
-
-	const isImage = (value: OnFileReadParams | string) => {
-		return dataUrl(value)?.startsWith('data:image/');
-	};
+	const disabled = uiSchema?.['ui:options']?.disabled as boolean | undefined;
```
Comment (JSReds) [inline src/components/Form/Widgets/FileWidget.tsx:115]: To ensure clarity, should we add a `disabledReasons` property and a tooltip to explain to the user why drag-and-drop is disabled?

Also, the cast here is not necessary

```suggestion
	const disabled = uiSchema?.['ui:options']?.disabled;
```

File: src/components/Form/Widgets/FileWidget.tsx [hunk 3]
```diff
@@ -1,248 +1,429 @@
...
+			if (multiple) {
+				setFiles([
+					...files,
+					...acceptedFiles,
+					...rejectedFiles.map(
+						(rejectedFile) =>
+							({
+								...rejectedFile.file,
+								// for some reason, spreading rejectedFile.file does not spread these properties
+								name: rejectedFile.file.name,
```
Comment (JSReds) [inline src/components/Form/Widgets/FileWidget.tsx:192]: It's expected as the interfaces are different.

```
type AcceptedFile = File & {
    loadingPercentage?: number;
    base64Data?: string;
    errors?: Array<{ code: ErrorCode | undefined; message: string }>;
};

interface FileRejection {
    file: File;
    errors: FileError[];
}
```

The issue isn't with the spread operator failing to spread properties;
rather, these properties simply don't exist on the object.

Additionally, why are we attempting to display the file if it was rejected?
If the file is rejected, I would expect the widget not load the file but instead only display the error.

File: src/components/Form/Widgets/FileWidget.tsx [hunk 4]
```diff
@@ -1,248 +1,429 @@
...
+			} else {
+				setFiles(
+					!!acceptedFiles.length
+						? [acceptedFiles[0]]
+						: !!rejectedFiles.length
+						? [
+								{
+									...rejectedFiles[0].file,
+									// for some reason, spreading rejectedFile.file does not spread these properties
+									name: rejectedFiles[0].file.name,
```
Comment (JSReds) [inline src/components/Form/Widgets/FileWidget.tsx:208]: expected as above

File: src/index.tsx [hunk 5]
```diff
@@ -24,12 +24,7 @@ export type {
 	UiSchema,
 	FormValidation,
 } from './components/Form';
-export type {
-	FileWidget,
```
Comment (JSReds) [inline src/index.tsx:28]: I recommend keeping the export on the fileWidget, as it might be needed in the future.

File: src/components/Form/Widgets/FileWidget.tsx [hunk 6]
```diff
@@ -1,248 +1,429 @@
...
+};
+
+const dragStyle: SxProps = {
+	borderColor: designTokens.color.border.accent.strong.value,
+	borderWidth: '3px',
+	backgroundColor: designTokens.color.bg.accent.value,
+};
+
+// These are just the ones we use in the dashboard at the moment
+// (excluding XML which is a pro icon, for which we just default to faFile)
```
Comment (drskullster) [inline src/components/Form/Widgets/FileWidget.tsx:49]: Maybe add a TODO to add the XML icon once we have pro icons in ui-shared?

File: src/components/Form/Widgets/FileWidget.tsx [hunk 7]
```diff
@@ -1,248 +1,429 @@
...
+		case 'image/jpg':
+		case 'image/jpeg':
+			return faFileImage;
+		default:
+			return faFile;
+	}
+};
+
+const fileSizeToReadable = (size: number) => {
+	const i = Math.floor(Math.log(size) / Math.log(1000));
```
Comment (drskullster) [inline src/components/Form/Widgets/FileWidget.tsx:63]: Shouldn't it be `1024`? 🤔

File: src/components/Form/Widgets/FileWidget.tsx (hunk 3 above)
Comment (myarmolinsky) [reply src/components/Form/Widgets/FileWidget.tsx:192]: The idea is to show the user a list of the files they tried to upload and the reasons the uploads failed per file. I will send the design in zulip, seeing it should be helpful

File: src/components/Form/Widgets/FileWidget.tsx (hunk 7 above)
Comment (myarmolinsky) [reply src/components/Form/Widgets/FileWidget.tsx:63]: 1024 is binary, I'm pretty sure react dropzone uses decimal

Comment (otaviojacobi) [reply src/components/Form/Widgets/FileWidget.tsx:63]: I also think it should be 1024. This is still decimal, just breaking up on 1024bytes at a time

File: package-lock.json [hunk 8]
```diff
@@ -9,7 +9,7 @@
       "version": "5.8.2",
       "license": "ISC",
       "dependencies": {
-        "@balena/design-tokens": "^0.7.2",
```
Comment (JSReds) [inline package-lock.json:12]: This file contains numerous changes that are unrelated to the module installation. Could you please install only the specific module without running npm install?

File: src/components/Form/Widgets/FileWidget.tsx [hunk 9]
```diff
@@ -1,248 +1,478 @@
...
+		'small';
+	const uploadIcon = uiSchema?.['ui:options']?.uploadIcon as
+		| IconDefinition
+		| undefined;
+	const descriptiveText = uiSchema?.['ui:options']?.descriptiveText as
+		| string
+		| undefined;
+	const hideUploadedFiles = uiSchema?.['ui:options']?.hideUploadedFiles as
+		| boolean
+		| undefined;
```
Comment (JSReds) [inline src/components/Form/Widgets/FileWidget.tsx:147]: what about something like this to avoid all these type casts ? (feel free to update it, just a basic idea)

```
type UIOptions = {
	accept?: Accept;
	maxSize?: number;
	disabled?: boolean;
	disabledReason?: string;
	browseResourceName?: string;
	variant?: 'large' | 'small';
	uploadIcon?: IconDefinition;
	descriptiveText?: string;
	hideUploadedFiles?: boolean;
};

const options = uiSchema?.['ui:options'] as UIOptions | undefined;

const accept = options?.accept;
const maxSize = options?.maxSize;
const disabled = options?.disabled;
const disabledReason = options?.disabledReason;
const browseResourceName = options?.browseResourceName;
const isSmall = options?.variant === 'small';
const uploadIcon = options?.uploadIcon;
const descriptiveText = options?.descriptiveText;
const hideUploadedFiles = options?.hideUploadedFiles;
```

File: src/components/Form/Widgets/FileWidget.tsx [hunk 10]
```diff
@@ -1,248 +1,478 @@
...
+							{errorFiles
+								.map((errorFile) => errorFile.errors)
+								.flat()
+								.map((error) =>
+									getErrorMessage(
+										error.code,
+										error.message,
+										maxSize ?? Infinity,
+									),
+								)}
```
Comment (JSReds) [inline src/components/Form/Widgets/FileWidget.tsx:340]:
```suggestion
.flatMap((errorFile) => errorFile.errors)
.map((error) =>
	getErrorMessage(
		error.code,
		error.message,
		maxSize ?? Infinity,
	),
)}
```

File: src/components/Form/Widgets/FileWidget.tsx [hunk 11]
```diff
@@ -1,248 +1,478 @@
...
+											</Typography>
+											<Typography color={designTokens.color.text.subtle.value}>
+												{fileSizeToReadable(file.size)}
+											</Typography>
+										</Stack>
+										<IconButtonWithTracking
+											eventName="FileWidget: Delete uploaded file"
+											aria-label="Delete uploaded file"
+											onClick={() => removeFile(index)}
+											// TODO: Jon please confirm
```
Comment (JSReds) [inline src/components/Form/Widgets/FileWidget.tsx:410]: I would add it as a PR comment more then a TODO

File: src/components/Form/Widgets/FileWidget.tsx (hunk 9 above)
Comment (myarmolinsky) [reply src/components/Form/Widgets/FileWidget.tsx:147]: I was thinking about the same, don't remember why I dropped the idea, let me do it

File: src/components/Form/Widgets/FileWidget.tsx [hunk 12]
```diff
@@ -1,248 +1,477 @@
...
+											</Typography>
+											<Typography color={designTokens.color.text.subtle.value}>
+												{fileSizeToReadable(file.size)}
+											</Typography>
+										</Stack>
+										<IconButtonWithTracking
+											eventName="FileWidget: Delete uploaded file"
+											aria-label="Delete uploaded file"
+											onClick={() => removeFile(index)}
+											color="secondary"
```
Comment (myarmolinsky) [inline src/components/Form/Widgets/FileWidget.tsx:410]: @drskullster please confirm we should indeed be changing the color of this button to "secondary"

Comment (drskullster) [reply src/components/Form/Widgets/FileWidget.tsx:410]: it should be the default color @myarmolinsky, but I realize it's not the same in my mockup than in the design system. I say just remove the `color` prop and roll with that, we'll adapt in a future PR if needs be.

**Review by JSReds — APPROVED** [review]
